warehouse/keyword-common/src/main/java/datawave/util/keyword/YakeKeywordExtractor.java

```diff
@@ -0,0 +1,718 @@
+
+    /**
+     * Main entrypoint. Extract keywords from the string provided.
+     *
+     * @param input
+     *            a string to process.
+     * @return a map of keywords to scores, ranked from highest to lowest score.
+     */
+    public LinkedHashMap<String,Double> extractKeywords(String input) {
+        // trim the input to the max content length if configured.
+        final String trimmedInput = (maxContentLength > 0 && input.length() > maxContentLength) ? input.substring(0, maxContentLength) : input;
+        // remove newlines because they are interpreted as sentence breaks.
+        final String normalizedInput = trimmedInput.replaceAll("[\\r\\n]+", " ");
+        final List<String> sentenceList = breakSentences(normalizedInput);
+        final List<List<String>> tokenizedSentences = tokenizeSentences(sentenceList);
+        return extractKeywords(tokenizedSentences);
+    }
+
+    /**
+     * Extract keywords from a collection of tokenized sentences. Ties each of the processing steps together into a complete workflow.
+     *
+     * @param tokenizedSentences
+     *            a list of lists; sentences that are already broken into tokens.
+     * @return a map of keywords to scores.
+     */
+    protected LinkedHashMap<String,Double> extractKeywords(List<List<String>> tokenizedSentences) {
+        List<TokenValue> basicStats = calculateBasicStats(tokenizedSentences);
+        List<List<String>> sentences = downCaseSentences(tokenizedSentences);
+        final List<List<String>> ngrams = calculateNGrams(sentences, minNGrams, maxNGrams);
+        Map<String,Map<String,Integer>> coLeft = buildCoOccurrence(ngrams, true);
+        Map<String,Map<String,Integer>> coRight = buildCoOccurrence(ngrams, false);
+        List<TaggedToken> taggedSentences = assignTags(basicStats);
+        List<YakeToken> tokens = calculateTokenScores(basicStats, taggedSentences, coLeft, coRight);
+        Map<String,Long> candidateKeywords = calculateCandidateKeywords(taggedSentences);
+        return calculateKeywords(candidateKeywords, tokens);
+    }
+
+    /**
+     * Given a set of tokenized sentences, generate a list of tuples containing (word, sentence id) pairs. The word is unchanged from its original form.
+     *
+     * @param tokenizedSentences
+     *            the list of tokenized sentences
+     * @return tuple of (word, sentence id pairs).
+     */
+    protected static List<TokenValue> calculateBasicStats(List<List<String>> tokenizedSentences) {
+        final List<TokenValue> resultsFlattenedIndexed = new ArrayList<>();
+
+        //@formatter:off
+        IntStream.range(0, tokenizedSentences.size())
+                .mapToObj(i ->
+                        tokenizedSentences.get(i)
+                                .stream()
+                                .map(t -> new TokenValue(t, i)))
+                .reduce(Stream.empty(), Stream::concat)
+                .collect(Collectors.toCollection(() -> resultsFlattenedIndexed));
```

Comment:
this is fairly complicated for what it does. This is a little simpler if you want to use streams

```suggestion
        AtomicInteger sentenceIndex = new AtomicInteger(0);
        return tokenizedSentences.stream()
                .flatMap(sentence -> {
                    int currentIndex = sentenceIndex.getAndIncrement();
                    return sentence.stream().map(token -> new TokenValue(token, currentIndex));
                })
                .collect(Collectors.toCollection(ArrayList::new));
```

Reply:
Yes, too complicated indeed. This is even simpler:
```java
        final List<TokenValue> resultsFlattenedIndexed = new ArrayList<>();
        for (int i=0; i < tokenizedSentences.size(); i++) {
            for (String token: tokenizedSentences.get(i)) {
                resultsFlattenedIndexed.add(new TokenValue(token, i));
            }
        }
        return resultsFlattenedIndexed;
```